File: go/internal/services/anilist/notify.go
package anilist

import (
	"context"
	"fmt"
	"log"
	"sync"
	"time"

	"discord-anime-bot/internal/services/redis"
	"discord-anime-bot/internal/types"

	"github.com/bwmarrin/discordgo"
)

// GetNextEpisode gets the next airing episode for an anime
func GetNextEpisode(animeID int) (*types.NextAiringEpisode, error) {
	anime, err := GetAnimeByID(animeID)
	if err != nil {
		return nil, err
	}

	return anime.NextAiringEpisode, nil
}

// notificationTimer holds a notification with its timer
type notificationTimer struct {
	Entry      *types.NotificationEntry
	Timer      *time.Timer
	CancelFunc context.CancelFunc
}

// NotificationService handles episode notifications
type NotificationService struct {
	notifications map[string]*notificationTimer
	session       *discordgo.Session
	mu            sync.RWMutex
}

// NewNotificationService creates a new notification service
func NewNotificationService(session *discordgo.Session) *NotificationService {
	service := &NotificationService{
		notifications: make(map[string]*notificationTimer),
		session:       session,
	}

	// Load existing notifications
	service.loadNotifications()

	return service
}

// loadNotifications loads notifications from Redis and reschedules them
func (ns *NotificationService) loadNotifications() {
	ctx := context.Background()

	// Get all notification keys from Redis
	notificationKeys, err := redis.Keys(ctx, "notification:*")
	if err != nil {
		log.Printf("Error fetching notification keys from Redis: %v", err)
		return
	}

	if len(notificationKeys) == 0 {
		log.Println("No notifications found in Redis")
		return
	}

	now := time.Now()
	loadedCount := 0

	for _, redisKey := range notificationKeys {
		var persistedNotification types.PersistedNotification
		err := redis.Get(ctx, redisKey, &persistedNotification)
		if err != nil {
			log.Printf("Error getting notification %s from Redis: %v", redisKey, err)
			continue
		}

		// Skip expired notifications
		airingTime := time.Unix(persistedNotification.AiringAt/1000, 0)
		if airingTime.Before(now) {
			// Remove expired notification
			if err := redis.Delete(ctx, redisKey); err != nil {
				log.Printf("Error deleting expired notification %s from Redis: %v", redisKey, err)
			}
			continue
		}

		notification := types.NotificationEntry{
			AnimeID:   persistedNotification.AnimeID,
			ChannelID: persistedNotification.ChannelID,
			UserID:    persistedNotification.UserID,
			AiringAt:  persistedNotification.AiringAt / 1000, // Convert to seconds
			Episode:   persistedNotification.Episode,
		}

		ns.scheduleNotificationInternal(&notification)
		loadedCount++
	}

	log.Printf("Loaded %d active notifications from Redis", loadedCount)
}

// saveNotificationToRedis saves a single notification to Redis
func (ns *NotificationService) saveNotificationToRedis(notificationKey string, entry *types.NotificationEntry) error {
	ctx := context.Background()
	redisKey := "notification:" + notificationKey

	persistedEntry := types.PersistedNotification{
		AnimeID:   entry.AnimeID,
		ChannelID: entry.ChannelID,
		UserID:    entry.UserID,
		AiringAt:  entry.AiringAt * 1000, // Convert to milliseconds for consistency
		Episode:   entry.Episode,
	}

	// Calculate TTL based on airing time (with buffer)
	airingTime := time.Unix(entry.AiringAt, 0)
	ttl := time.Until(airingTime) + time.Hour // 1 hour buffer
	if ttl <= 0 {
		ttl = time.Minute // Minimum 1 minute TTL
	}

	err := redis.Set(ctx, redisKey, persistedEntry, ttl)
	if err != nil {
		return fmt.Errorf("failed to save notification to Redis: %w", err)
	}

	log.Printf("Saved notification %s to Redis", notificationKey)
	return nil
}

// removeNotificationFromRedis removes a notification from Redis
func (ns *NotificationService) removeNotificationFromRedis(notificationKey string) error {
	ctx := context.Background()
	redisKey := "notification:" + notificationKey

	err := redis.Delete(ctx, redisKey)
	if err != nil {
		return fmt.Errorf("failed to remove notification from Redis: %w", err)
	}

	log.Printf("Removed notification %s from Redis", notificationKey)
	return nil
}

// createNotificationKey creates a unique key for a notification
func createNotificationKey(animeID int, channelID, userID string) string {
	return fmt.Sprintf("%d-%s-%s", animeID, channelID, userID)
}

// scheduleNotificationInternal sets up a timer for the notification (without locking)
func (ns *NotificationService) scheduleNotificationInternal(entry *types.NotificationEntry) {
	notificationKey := createNotificationKey(entry.AnimeID, entry.ChannelID, entry.UserID)

	// Calculate delay until airing time
	airingTime := time.Unix(entry.AiringAt, 0)
	delay := time.Until(airingTime)
	if delay <= 0 {
		log.Printf("Warning: Notification for anime %d is in the past, skipping", entry.AnimeID)
		return
	}

	ctx, cancel := context.WithCancel(context.Background())

	timer := time.AfterFunc(delay, func() {
		select {
		case <-ctx.Done():
			return
		default:
			ns.sendNotification(entry)
			// Remove the notification after sending
			ns.mu.Lock()
			delete(ns.notifications, notificationKey)
			ns.mu.Unlock()
			// Remove from Redis as well
			if err := ns.removeNotificationFromRedis(notificationKey); err != nil {
				log.Printf("Error removing notification from Redis: %v", err)
			}
		}
	})

	ns.notifications[notificationKey] = &notificationTimer{
		Entry:      entry,
		Timer:      timer,
		CancelFunc: cancel,
	}

	log.Printf("Scheduled notification for anime %d in %v", entry.AnimeID, delay)
}

// sendNotification sends the Discord notification
func (ns *NotificationService) sendNotification(entry *types.NotificationEntry) {
	// Get anime details for the notification
	anime, err := GetAnimeByID(entry.AnimeID)
	if err != nil {
		log.Printf("Error getting anime details for notification: %v", err)
		return
	}

	title := anime.Title.Romaji
	if anime.Title.English != nil && *anime.Title.English != "" {
		title = *anime.Title.English
	}

	embed := &discordgo.MessageEmbed{
		Title:       "Episode Alert!",
		Description: fmt.Sprintf("**Episode %d** of **%s** is now airing!", entry.Episode, title),
		Color:       0x00FF00,
		Thumbnail: &discordgo.MessageEmbedThumbnail{
			URL: anime.CoverImage.Large,
		},
		Footer: &discordgo.MessageEmbedFooter{
			Text: "Set more notifications with /anime notify add",
		},
		Timestamp: time.Now().Format(time.RFC3339),
	}

	content := fmt.Sprintf("<@%s>", entry.UserID)

	_, err = ns.session.ChannelMessageSendComplex(entry.ChannelID, &discordgo.MessageSend{
		Content: content,
		Embeds:  []*discordgo.MessageEmbed{embed},
	})

	if err != nil {
		log.Printf("Error sending notification: %v", err)
	} else {
		log.Printf("Sent notification for anime %s episode %d to user %s", title, entry.Episode, entry.UserID)
	}
}

// AddNotification adds a new episode notification
func (ns *NotificationService) AddNotification(animeID int, channelID, userID string, airingAt time.Time, episode int) error {
	notificationKey := createNotificationKey(animeID, channelID, userID)

	ns.mu.Lock()
	defer ns.mu.Unlock()

	// Check if notification already exists
	if existing, exists := ns.notifications[notificationKey]; exists {
		existing.Timer.Stop()
		existing.CancelFunc()
		delete(ns.notifications, notificationKey)
	}

	entry := &types.NotificationEntry{
		AnimeID:   animeID,
		ChannelID: channelID,
		UserID:    userID,
		AiringAt:  airingAt.Unix(),
		Episode:   episode,
	}

	log.Printf("Adding notification for anime %d, episode %d", animeID, episode)

	// Schedule the notification (internal method that doesn't acquire locks)
	ns.scheduleNotificationInternal(entry)

	// Save to Redis
	return ns.saveNotificationToRedis(notificationKey, entry)
}

// RemoveNotification removes a notification for a specific anime and user
func (ns *NotificationService) RemoveNotification(animeID int, channelID, userID string) error {
	notificationKey := createNotificationKey(animeID, channelID, userID)

	ns.mu.Lock()
	defer ns.mu.Unlock()

	timer, exists := ns.notifications[notificationKey]
	if !exists {
		return fmt.Errorf("notification not found for anime %d and user %s", animeID, userID)
	}

	timer.Timer.Stop()
	timer.CancelFunc()
	delete(ns.notifications, notificationKey)

	// Remove from Redis
	return ns.removeNotificationFromRedis(notificationKey)
}

// GetUserNotifications returns all notifications for a specific user
func (ns *NotificationService) GetUserNotifications(userID string) []*types.NotificationEntry {
	ns.mu.RLock()
	defer ns.mu.RUnlock()

	var notifications []*types.NotificationEntry
	for _, timer := range ns.notifications {
		if timer.Entry.UserID == userID {
			notifications = append(notifications, timer.Entry)
		}
	}

	return notifications
}

// Cleanup stops all timers
func (ns *NotificationService) Cleanup() {
	ns.mu.Lock()
	defer ns.mu.Unlock()

	log.Println("Cleaning up notification service...")

	for _, timer := range ns.notifications {
		timer.Timer.Stop()
		timer.CancelFunc()
	}

	log.Println("Notification service cleanup complete")
}

File: src/services/anime/notify.ts
import { Client, TextChannel } from 'discord.js'
import { getAnimeById } from './next'
import type { NotificationEntry } from '@/types/anilist'
import { redisCache } from '@/services/redis/cache'

let notifications: Map<string, NotificationEntry> = new Map()
let client: Client | null = null

/**
 * Set the Discord client and load notifications on startup
 */
export function setClient(newClient: Client) {
  client = newClient
  // Load existing notifications on startup
  loadNotifications()
}

/**
 * Load notifications from Redis and reschedule them
 */
async function loadNotifications() {
  try {
    // Get all notification keys from Redis
    const notificationKeys = await redisCache.getKeys('notification:*')

    if (notificationKeys.length === 0) {
      console.log('No notifications found in Redis')
      return
    }

    const now = Date.now()
    let loadedCount = 0

    for (const redisKey of notificationKeys) {
      const persistedNotification = await redisCache.get<NotificationEntry>(redisKey, true)

      if (!persistedNotification) continue

      // Skip expired notifications
      if (persistedNotification.airingAt <= now) {
        await redisCache.delete(redisKey)
        continue
      }

      const notificationKey = `${persistedNotification.animeId}-${persistedNotification.channelId}-${persistedNotification.userId}`
      const timeUntilAiring = persistedNotification.airingAt - now

      const entry: NotificationEntry = {
        ...persistedNotification,
        timeoutId: setTimeout(() => {
          sendNotification(entry)
        }, timeUntilAiring)
      }

      notifications.set(notificationKey, entry)
      loadedCount++
    }

    console.log(`Loaded ${loadedCount} active notifications from Redis`)
  } catch (error) {
    console.error('Error loading notifications from Redis:', error)
  }
}

/**
 * Save a single notification to Redis
 */
async function saveNotification(notificationKey: string, entry: NotificationEntry): Promise<void> {
  try {
    const redisKey = `notification:${notificationKey}`
    const persistedEntry = {
      animeId: entry.animeId,
      channelId: entry.channelId,
      userId: entry.userId,
      airingAt: entry.airingAt,
      episode: entry.episode
    }

    // Calculate TTL based on airing time (with some buffer)
    const now = Date.now()
    const ttlSeconds = Math.max(Math.ceil((entry.airingAt - now) / 1000) + 3600, 60) // At least 1 minute TTL

    await redisCache.set(redisKey, persistedEntry, ttlSeconds)
    console.log(`[Save] Successfully saved notification ${notificationKey} to Redis`)
  } catch (error) {
    console.error(`[Save] Error saving notification ${notificationKey} to Redis:`, error)
  }
}

/**
 * Remove a notification from Redis
 */
async function removeNotificationFromRedis(notificationKey: string): Promise<void> {
  try {
    const redisKey = `notification:${notificationKey}`
    await redisCache.delete(redisKey)
    console.log(`[Remove] Successfully removed notification ${notificationKey} from Redis`)
  } catch (error) {
    console.error(`[Remove] Error removing notification ${notificationKey} from Redis:`, error)
  }
}

/**
 * Add a notification for an anime episode
 */
export async function addNotification(
  animeId: number,
  channelId: string,
  userId: string
): Promise<{ success: boolean; message: string; airingDate?: Date }> {
  if (!client) {
    return { success: false, message: 'Bot client not initialized' }
  }
  try {
    const anime = await getAnimeById(animeId)

    if (!anime) {
      return { success: false, message: `No anime found with ID ${animeId}` }
    }

    if (!anime.nextAiringEpisode) {
      return {
        success: false,
        message: `No upcoming episodes scheduled for ${anime.title.english || anime.title.romaji}`
      }
    }

    if (anime.status === 'FINISHED') {
      return { success: false, message: `This anime has finished airing` }
    }

    if (anime.status === 'CANCELLED') {
      return { success: false, message: `This anime has been cancelled` }
    }

    const airingAt = anime.nextAiringEpisode.airingAt * 1000 // Convert to milliseconds
    const now = Date.now()

    if (airingAt <= now) {
      return { success: false, message: 'This episode has already aired' }
    }

    const notificationKey = `${animeId}-${channelId}-${userId}`

    // Check if notification already exists
    if (notifications.has(notificationKey)) {
      return {
        success: false,
        message: `You already have a notification set for ${anime.title.english || anime.title.romaji}`
      }
    }

    // Remove any other notifications for this anime by this user (in case they exist in other channels)
    const keysToRemove: string[] = []
    for (const [key, entry] of notifications.entries()) {
      if (entry.animeId === animeId && entry.userId === userId) {
        keysToRemove.push(key)
      }
    }

    // Remove the found notifications (without saving each time to avoid race condition)
    for (const key of keysToRemove) {
      removeNotificationInternal(key)
    }

    const airingDate = new Date(airingAt)
    const timeUntilAiring = airingAt - now

    const entry: NotificationEntry = {
      animeId,
      channelId,
      userId,
      airingAt,
      episode: anime.nextAiringEpisode.episode
    }

    // Schedule the notification
    entry.timeoutId = setTimeout(() => {
      sendNotification(entry)
    }, timeUntilAiring)

    notifications.set(notificationKey, entry)

    // Save to Redis
    await saveNotification(notificationKey, entry)

    return {
      success: true,
      message: `Notification set for ${anime.title.english || anime.title.romaji} Episode ${anime.nextAiringEpisode.episode}`,
      airingDate
    }
  } catch (error) {
    console.error('Error adding notification:', error)
    return { success: false, message: 'An error occurred while setting up the notification' }
  }
}

/**
 * Remove a notification (internal method without saving)
 */
function removeNotificationInternal(notificationKey: string): boolean {
  const entry = notifications.get(notificationKey)
  if (entry?.timeoutId) {
    clearTimeout(entry.timeoutId)
  }
  return notifications.delete(notificationKey)
}

/**
 * Remove a notification
 */
export async function removeNotification(notificationKey: string): Promise<boolean> {
  const removed = removeNotificationInternal(notificationKey)
  if (removed) {
    await removeNotificationFromRedis(notificationKey)
  }
  return removed
}

/**
 * Remove notifications for a specific user and anime
 */
export async function removeUserNotification(
  animeId: number,
  channelId: string,
  userId: string
): Promise<boolean> {
  const notificationKey = `${animeId}-${channelId}-${userId}`
  return await removeNotification(notificationKey)
}

/**
 * Send the notification when episode airs
 */
async function sendNotification(entry: NotificationEntry) {
  if (!client) return
  try {
    const channel = (await client.channels.fetch(entry.channelId)) as TextChannel
    if (!channel) return
    const anime = await getAnimeById(entry.animeId)
    if (!anime) return
    const title = anime.title.english || anime.title.romaji
    const message = `<@${entry.userId}> Episode ${entry.episode} of **${title}** has just aired!\n\n[AniList Details](${anime.siteUrl})`
    await channel.send(message)
    // Remove the notification after sending
    const notificationKey = `${entry.animeId}-${entry.channelId}-${entry.userId}`
    notifications.delete(notificationKey)
    await removeNotificationFromRedis(notificationKey)
  } catch (error) {
    console.error('Error sending notification:', error)
  }
}

/**
 * Get all active notifications for a user in a specific channel
 */
export function getUserNotifications(userId: string, channelId?: string): NotificationEntry[] {
  return Array.from(notifications.values()).filter(
    (entry) => entry.userId === userId && (!channelId || entry.channelId === channelId)
  )
}

/**
 * Clean up expired notifications (for safety)
 */
export async function cleanup() {
  const now = Date.now()
  const keysToRemove: string[] = []

  for (const [key, entry] of notifications.entries()) {
    if (entry.airingAt <= now) {
      if (entry.timeoutId) {
        clearTimeout(entry.timeoutId)
      }
      notifications.delete(key)
      keysToRemove.push(key)
    }
  }

  // Remove expired notifications from Redis
  for (const key of keysToRemove) {
    await removeNotificationFromRedis(key)
  }

  if (keysToRemove.length > 0) {
    console.log(`[Cleanup] Removed ${keysToRemove.length} expired notifications`)
  }
}
